aten/src/ATen/core/IList.h

```diff
@@ -0,0 +1,512 @@
+
+  IListIterator(boxed_iterator_type boxed) : tag_(IListTag::Boxed) {
+    payload_.boxed_iterator = boxed;
+  }
+
+  IListIterator(unboxed_iterator_type unboxed) : tag_(IListTag::Unboxed) {
+    payload_.unboxed_iterator = unboxed;
+  }
+
+  detail::IListConstRef<T> operator*() const {
+    TORCH_ILIST_UNWRAP(tag_, { return ImplT::iterator_get(this_); });
+  }
+
+  IListIterator& operator++() {
+    TORCH_ILIST_UNWRAP(tag_, { ++this_; });
+    return *this;
+  }
+
+  IListIterator operator++(int) {
+    auto old = *this;
+    TORCH_ILIST_UNWRAP(tag_, { ++this_; });
+    return old;
+  }
+
+  IListIterator& operator--() {
+    TORCH_ILIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  IListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ILIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const IListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ILIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const IListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  union Payload {
+    boxed_iterator_type boxed_iterator;
+    unboxed_iterator_type unboxed_iterator;
+    void* _init_ptr;
+    Payload() : _init_ptr(nullptr) {}
+    ~Payload() {}
+  };
+
+  Payload payload_;
+  IListTag tag_;
+};
+
+/*
+ * Wrapper around different API containers (e.g. boxed and unboxed).
+ *
+ * What is it?
+ * ===========
+ * It is a tagged union of both boxed and unboxed API containers.
+ * Working implementations:
+ *
+ * - `IList<at::Tensor>`
+ * - `IList<at::OptionalTensorRef>`
+ *
+ * Note that `ITensorList` is a view type. Meaning that it won't own the
+ * tensors it holds. If you need it to last longer, make sure that there is
+ * actually a non-temporary list of tensors (e.g. `vector<Tensor>`) that owns
+ * them and outlives the `ITensorList` instance.
+ *
+ * What is this for?
+ * =================
+ * Historically, PyTorch has maintained 2 different APIs: the unboxed
+ * (called from C++ API and Python eager mode) and boxed APIs (called
+ * from the TorchScript JIT, mobile interpreter, and boxed fallbacks).
+ *
+ * Calling unboxed kernels from the boxed "world" and vice-versa may
+ * result in non-negligible overhead. Lists are one of those types:
+ *
+ * - Boxed world: `c10::List`
+ * - Unboxed world: `c10::ArrayRef`
+ *
+ * In this context, `c10::IList` solves this problem by wrapping those
+ * 2 container types, so that we don't need to convert from one to
+ * the other.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ *
+ * What does it do?
+ * ================
+ * This container wraps around the different tagged containers
+ * (currently, only boxed and unboxed), without incurring in extra
+ * overhead for converting from one to another. It does so while
+ * exposing usual container methods, which dispatch to corresponding
+ * implementations.
```

Comment:
Maybe it would be good to say that the intended use of this interface is iterating over it once (and not repeatedly accessing elements from the container).